API: `sharp.overlayWith` (alpha compositing)

gasi opened the pull request:
Adds `sharp.overlaysWith` for compositing an image on top of another image while handling alpha channels correctly.

I left all commits intact for better understanding but feel free to merge as one squash commit, @lovell.

This is now ready for review.

**Tests and usage examples:** `test/unit/overlay.js`

**TODO**
- [x] ~~Test with non-PNGs.~~ Non-4-channel images are not supported.
- [x] ~~Add support for specifying `top` and `left` position as in `sharp.overlaysWith(filename, top, left)`. This might be useful for watermarking.~~ Punted.
- [ ] Benchmark.
- [ ] Investigate (slight) discrepancy to ImageMagick `convert -composite` output. Not visible to human eye (at least not mine :wink:) but detectable by `compare …`.

**References**
- http://en.wikipedia.org/wiki/Alpha_compositing#Alpha_blending
- https://github.com/jcupitt/ruby-vips/issues/28#issuecomment-9014826
- https://github.com/gasi/sharp/pull/2

See: #97.

gasi commented on CONTRIBUTING.md:
```diff
@@ -30,6 +30,8 @@ Please select the `master` branch as the destination for your Pull Request so yo
 
 Please squash your changes into a single commit using a command like `git rebase -i upstream/master`.
 
+To test C++ changes, you can compile the module using `npm run compile`.
```
This was useful to me and hope it will be to others :smile:

lovell replied:
This will compile but not test the changes. I usually use `npm i` to do both.

gasi replied:
Oh, you’re right, I didn’t know `npm install` runs `node-gyp` on the package itself. I always tried `npm rebuild` but that only recompiles dependencies. However, you said it ‘will […] test the changes’. This doesn’t happen for me (see below). Am I missing something?

```
npm i

> sharp@0.10.0 install /Users/dani/workspace/sharp
> node-gyp rebuild

  CXX(target) Release/obj.target/sharp/src/common.o
  CC(target) Release/obj.target/sharp/src/composite.o
  CXX(target) Release/obj.target/sharp/src/utilities.o
  CXX(target) Release/obj.target/sharp/src/metadata.o
  CXX(target) Release/obj.target/sharp/src/resize.o
  CXX(target) Release/obj.target/sharp/src/sharp.o
  SOLINK_MODULE(target) Release/sharp.node
  SOLINK_MODULE(target) Release/sharp.node: Finished
```

lovell replied:
Sorry, my mistake. Judging by the command line history `npm i && npm t` will do both.

gasi replied:
No worries, I documented `npm install` and `npm test` :+1:

gasi commented on index.js:
```diff
@@ -60,6 +60,9 @@ var Sharp = function(input) {
     greyscale: false,
     normalize: 0,
+    // overlay
+    hasOverlay: false,
```
@lovell: If you show me how to test check if a string is empty in C++, we might be able to get rid of `hasOverlay`. On the other hand, having this flag might be more understandable instead of just relying on empty string `overlayPath`.

lovell replied:
std::string provides an [empty](http://www.cplusplus.com/reference/string/string/empty/) method.

gasi replied:
:+1: That worked, thanks.

gasi commented on package.json:
```diff
@@ -48,13 +49,14 @@
     "node-cpplint": "^0.4.0",
-    "rimraf": "^2.3.2",
-    "async": "^0.9.0"
+    "rimraf": "^2.3.2"
```
Alphabetized `devDependencies` for easier scanability and added `gm` to use `gm.compare` in automated tests.

gasi commented on src/composite.c:
```diff
@@ -0,0 +1,119 @@
+
+/* Composite images `src` and `dst`.
+ */
+int Composite(VipsObject *context, VipsImage *src, VipsImage *dst,
```
Used `PascalCase` for `Composite` (used to be `composite`) to match the rest of the codebase.

gasi replied:
I also tightened up whitespace to match this codebase’s style which is different from `libvips`.

lovell replied:
I try to keep the style JavaScript-like to help contributors from that arena :)

gasi replied:
:+1: to making it more familiar for JavaScript developers. Is `PascalCase` more common for C++ than `camelCase`?

lovell replied:
PascalCase for classes and camelCase for variable names seems to be near-standard across C++, Java and JavaScript. The Google C++ guidelines use PascalCase for (non-accessor) method names, but I suspect that may have been influenced by the prevalence of C++ for Windows development in the 90s. The Sun Java guidelines use camelCase for method names, which JavaScript seems to have inherited. 

The only conclusion I've been able to reach is that consistency within a code base is more important than consistency across code bases.

gasi replied:
> The Google C++ guidelines use PascalCase for (non-accessor) method names, but I suspect that may have been influenced by the prevalence of C++ for Windows development in the 90s.

Thanks for sharing, I didn’t know that. C# uses the same convention.



> The only conclusion I've been able to reach is that consistency within a code base is more important than consistency across code bases.

I wholeheartedly agree and tried to match my new code as closely as possible :+1:

gasi commented on src/composite.c:
```diff
@@ -0,0 +1,119 @@
+    return -1;
+
+  vips_object_local(context, srcRGB);
+  vips_object_local(context, dstRGB);
```
Thanks for the advice with `vips_object_local`. Please check that I made that change correctly.

lovell replied:
Thank you. You should also be able to replace the use of `vips_object_local_array` and `g_object_ref` with matching pairs of `VipsImage *example` and `vips_object_local(context, example)`.

gasi replied:
Do you mean using explicit (named) references for `t[0]`, `t[1]`, and `t[2]`? This suggestion originally came from John Cupitt in https://github.com/jcupitt/ruby-vips/issues/28#issuecomment-97049450. He used `t[…]` for all intermediate result and I thought it might be valuable to use named references where possible. What do you think? Sorry, if I misunderstood your suggestion.

lovell replied:
Yes, using named references is preferable to an array. This avoids the chance of being out-of-bounds, which C does not guard against.

gasi replied:
Thanks for your advice, I changed it to use individual references.

gasi commented on test/unit/overlay.js:
```diff
@@ -0,0 +1,113 @@
+// Helpers
+var tryDelete = function (file) {
+  try {
+    fs.unlinkSync(file);
```
I hope it’s ok I used a sync delete to save some typing.

lovell replied:
Perfect, tests are the right place to use simpler sync i/o.

gasi replied:
Thanks for `fixtures.assertSimilar` and your example use of `toBuffer`. This enabled me to kill the intermediate files altogether :smile:

gasi commented on test/unit/overlay.js:
```diff
@@ -0,0 +1,113 @@
+
+var compare = function (actual, expected, callback) {
+  // TODO: Allow for `tolerance` setting
+  gm.compare(actual, expected, function (error, result) {
```
I know this isn’t quite perceptual hashing #122 but it was helpful to me.

lovell commented:
Thanks again Daniel for all your work on this very useful and much-requested feature.

I'm not sure I want to go down the route of using *magick for image comparisons. My experience of managing CI environments warns me otherwise (I'm looking at you, [AppVeyor](https://ci.appveyor.com/project/lovell/sharp)) and, more importantly, I believe libvips can do this with minimum effort.

I'll spend some time this afternoon to see what I can do using libvips to get #122 off the ground.

lovell commented:
Commit c159948 on the `knife` branch adds a [fixtures.assertSimilar](https://github.com/lovell/sharp/blob/knife/test/fixtures/index.js#L78) method you can use instead of *magick's "compare".

By way of example, all of the tests for the [extract](https://github.com/lovell/sharp/blob/knife/test/unit/extract.js) method now employ it.

gasi commented:
> Commit c159948 on the knife branch adds a fixtures.assertSimilar method you can use instead of *magick's "compare".

Awesome! Thanks for providing this and I agree it’s more elegant without the GraphicsMagick dependency. However, I have two observations:

1. I found a case where the similarity test should fail: c9301c1fb5790c34adca1eb17fb1c1f953fdd76f
2. Isn’t it valuable for some operations like alpha compositing to do a comparison using no tolerance? If you agree, I could add that as parameter to `assertSimilar`.

gasi commented on .jshintrc:
```diff
@@ -4,6 +4,7 @@
   "maxparams": 4,
   "maxcomplexity": 13,
   "globals": {
+    "before": true,
```
Although I don’t use this anymore, I figure it might be valuable for the future :smile:

gasi commented on src/composite.c:
```diff
@@ -0,0 +1,119 @@
+  VipsImage *srcRGB;
+  VipsImage *dstRGB;
+  if (vips_extract_band(src, &srcRGB, 0, "n", NUM_COLOR_BANDS, NULL) ||
+      vips_extract_band(dst, &dstRGB, 0, "n", NUM_COLOR_BANDS, NULL))
```
This should probably use `sharp::HasAlpha` and it might be useful to add `sharp::GetAlpha` similar to `get_alpha` in https://github.com/jcupitt/ruby-vips/issues/28#issuecomment-9014826. 

Currently, we only have the need to composite RGBA PNGs for our purposes. Do you think we could land this PR like this or should we keep it open until any n-band images are correctly handled?

lovell replied:
To start with we can generate an explicit error if the image to be overlaid does not have 4 bands and an alpha channel.

gasi replied:
Done: 4c02c1d418fb4e7ca4208ef76c63019c930866f2 &  5edc24e77737b910aa81993d92f8cbe5f4ff9520

gasi commented on src/composite.h:
```diff
@@ -0,0 +1,17 @@
+/*
+  Composite images `src` and `dst`.
+*/
+int Composite(VipsObject *context, VipsImage *src, VipsImage *dst, VipsImage **out);
```
Getting used to your 140 line length limit. I usually operate at 80 out of habit :wink:

lovell replied:
People who use an 80 limit do so to leave room at the side of their monitor for Twitter.

gasi replied:
Haha, good one :stuck_out_tongue_winking_eye: On a more serious note: I do it because I like to do three-way merges on my laptop :+1:

gasi replied:
Then there is science :wink: http://baymard.com/blog/line-length-readability

gasi commented on src/resize.cc:
```diff
@@ -790,6 +792,33 @@ class ResizeWorker : public NanAsyncWorker {
+      ImageType overlayImageType = ImageType::UNKNOWN;
+      overlayImageType = DetermineImageType(baton->overlayPath.c_str());
+      if (overlayImageType != ImageType::UNKNOWN) {
+        overlayImage = InitImage(baton->overlayPath.c_str(), baton->accessMethod);
```
TODO: Handle non-transparent images more efficiently, e.g. using `vips_insert`.

gasi commented on test/unit/overlay.js:
```diff
@@ -0,0 +1,62 @@
+            if (error) return done(error);
+
+            // TODO: Remove once this fails
+            fixtures.assertSimilar(fixtures.expected('alpha-layer-012-low-alpha.png'), data, function (error, result) {
```
This `assertSimilar` should fail, but it doesn’t. Please see reason why it should fail below.

lovell replied:
The image fingerprinting method normalises image luminance so won't be able to detect all differences. It's worth trying your suggestion of exposing assertSimilar's threshold.

gasi replied:
Thanks, I added `threshold` option 23a151f20219a42d0615dfe0c851aee24607afae and it seems to have worked detecting that case using `threshold: 0`. However, I still wonder if it’d be valuable to have a identity check at an image level (ignoring metadata, etc) for testing subtle changes such as sharpening, opacity, etc.

gasi commented on test/unit/overlay.js:
```diff
@@ -0,0 +1,62 @@
+    sharp(fixtures.inputPngOverlayLayer0)
+      .overlayWith(fixtures.inputPngOverlayLayer1)
+      .toBuffer(function (error, data, info) {
+        if (error) return done(error);
```
Minor: I noticed you use `if (error) throw error;` in your unit tests but doesn’t mixing synchronous `throw` inside an async callback cause issues? I am fine doing whatever you prefer in terms of style. Just let me know.

lovell replied:
Yeah, good spot. I suspect `throw` still works as the `done` method will never get called but I definitely prefer your approach, thank you.

gasi replied:
:+1:

aseemk replied:
(Just FYI, Mocha catches synchronous errors during async test execution. That's necessary since every assertion library, including Node's own, throws errors synchronously.)

aseemk replied:
(Via `process.on('uncaughtException')`)

gasi replied:
Thanks for pointing this out, @aseemk. It makes sense in hindsight as the tests previously did work :smile:

gasi replied:
It’s probably still nicer to _throw_ async errors via callback, right?

aseemk replied:
I don't feel strongly, but think `throw`ing is maybe a bit better in test code — simpler, direct, and less room for bugs, all of which you particularly want in test code. But in this case, it's really up to @lovell. =)

lovell replied:
The mocha [async examples](http://mochajs.org/#asynchronous-code) contain both `throw err` and `done(err)` variants. Reducing the number of conditional branches in test code decreases the need to test-the-tests. In this case, both approaches introduce one `if/else` split. Promises (and generators) will take over the world soon so we can, ahem, *throw* the error handling part of this test code away at some point.

gasi replied:
:+1:

gasi commented on test/fixtures/index.js:
```diff
@@ -11,7 +11,7 @@ var getPath = function(filename) {
 // Generates a 64-bit-as-binary-string image fingerprint
 // Based on the dHash gradient method - see http://www.hackerfactor.com/blog/index.php?/archives/529-Kind-of-Like-That.html
-var fingerprint = function(image, done) {
+var fingerprint = function(image, callback) {
```
`callback` is a more common name than `done` in Node.js modules :+1:

gasi commented:
@lovell I made another pass on this today and it should be ready to merge given the restrictions we talked about (only RGBA images can be composited). I might be able to add more functionality at some point but for now, I think this might be already valuable for some people :smile:

lovell commented on src/composite.c:
```diff
@@ -0,0 +1,128 @@
+  vips_object_local(context, joined);
+
+  // Return a reference to the output image:
+  g_object_ref(joined);
```
I think this `g_object_ref` can be removed as `vips_object_local` will reference `joined`.

lovell commented:
Thanks again for all your work on this Daniel. Other than the possible reference leak, this is definitely ready to merge into the `knife` WIP branch. Are you able to squash to a single commit?

gasi commented:
> Other than the possible reference leak, this is definitely ready to merge into the knife WIP branch. Are you able to squash to a single commit?

Thanks, I simplified the `out` reference as you suggested and squashed everything into a single commit :smile: It’s been a pleasure collaborating with you and I hope I’ll be able to contribute more in the future.

gasi commented:
I forgot to add `overlayWith` documentation to `README`. Will do that later today.

lovell commented:
No worries on the docs. I intend to add the ability to overlay onto non-alpha images too and will update the README after that. Thanks again @gasi for all your work on this new feature!

gasi commented:
Thanks, @lovell, I appreciate it :+1:
